content/blog/2016/2016-08-10-rails-cd-with-pipeline.adoc

```diff
@@ -0,0 +1,358 @@
+link:https://github.com/rubycentral/cfp-app[cfp-app], a:
+
+[quote]
+____
+Ruby on Rails application that lets you manage your conference's call for
+proposal (CFP), program and schedule. It was written by Ruby Central to run the
+CFPs for RailsConf and RubyConf.
+____
+
+
+This isn't just some random Rails app I'm using for a pretty demo, this is
+actually the application we used to collect talk proposals for the
+"link:https://www.cloudbees.com/juc/agenda[Community Tracks]" at this year's
+link:http://jenkinsworld.com[Jenkins World]. For the most part, cfp-app is a
+fairly standard Rails application. It uses
+link:https://www.postgresql.org/[PostgreSQL] for its database,
+link:http://rspec.info[RSpec] for its tests and link:http://ruby-lang.org[Ruby
+2.3.x] as its runtime.
+
+
+[TIP]
+--
+If you prefer just to look at the code, skip straight to the
+link:https://github.com/rtyler/cfp-app/blob/aff11d97b460ca4b630b1fa4f669c226aeadd8ae/Jenkinsfile[Jenkinsfile].
+--
+
+
+=== Preparing the app
+
+For _most_ Rails applications there are little, if any, changes needed to enable
+continuous delivery with Jenkins. In the case of
+link:https://github.com/rubycentral/cfp-app[cfp-app], I added two gems to get
+the most optimal integration into Jenkins:
+
+. link:https://github.com/ci-reporter/ci_reporter[ci_reporter], for test report
+integration
+. link:https://github.com/presidentbeef/brakeman[brakeman], for security scanning.
+
+Adding these was simple, I just needed to update the `Gemfile` and the
+`Rakefile` in the root of the repository to contain:
+
+
+.Gemfile
+[source, ruby]
+----
+# .. snip ..
+group :test do
+  # RSpec, etc
+  gem 'ci_reporter'
+  gem 'ci_reporter_rspec'
+  gem "brakeman", :require => false
+end
+----
+
+.Rakefile
+[source, ruby]
+----
+# .. snip ..
+require 'ci/reporter/rake/rspec'
+# Make sure we setup ci_reporter before executing our RSpec examples
+task :spec => 'ci:setup:rspec'
+----
+
+
+=== Preparing Jenkins
+
+[[plugins]]
+
+The cfp-app project set up, next on the list is to ensure that Jenkins itself
+is ready. Generally I suggest running the link:/changelog-stable[latest LTS] of
+Jenkins; for this demonstration I used Jenkins 2.7.1 with the following
+plugins:
+
+* link:https://wiki.jenkins-ci.org/display/JENKINS/Pipeline+Plugin[Pipeline plugin]
+* link:https://wiki.jenkins-ci.org/display/JENKINS/Brakeman+Plugin[Brakeman plugin]
+* link:https://wiki.jenkins-ci.org/display/JENKINS/CloudBees+Docker+Pipeline+Plugin[CloudBees
+  Docker Pipeline plugin]
+
+I also used the
+link:https://wiki.jenkins-ci.org/display/JENKINS/GitHub+Organization+Folder+Plugin[GitHub
+Organization Folder plugin] to automatically create pipeline items in my
+Jenkins instance; that isn't required to the demo, but it's pretty cool to see
+repositories and branches with a `Jenkinsfile` automatically show up in
+Jenkins, so I recommend it!
+
+
+In addition to the <<plugins, plugins>> listed above, I also needed at least _one_
```

Comment:
This is an internal cross-reference to the `[[plugins]]` anchor above. The syntax is `<<anchor, displayText>>` basically. For some weird reason asciidoctor will render a single-word cross-reference as `[plugins]`.

Multiword cross-references don't display the brackets though :confounded: